pkg/dataplane/identity.go

```diff
@@ -69,14 +82,122 @@ func (c CredentialsObject) GetCredential(requestedResourceID string) (*azidentit
 
 // Get an AzIdentity credential for the given nested credential object
 // Clients can use the credential to get a token for the user-assigned identity
-func (n NestedCredentialsObject) GetCredential() (*azidentity.ClientCertificateCredential, error) {
+func (n *NestedCredentialsObject) GetCredential() (*azidentity.ClientCertificateCredential, error) {
 	if n.Values.ResourceID != nil {
 		return getClientCertificateCredential(n.Values, n.Cloud)
 	}
 
 	return nil, errResourceIDNotFound
 }
 
+func (n *NestedCredentialsObject) ReloadCredntialsOnChange() error {
+	if err := n.File.checkFileExists(); err != nil {
+		return err
+	}
+
+	err := n.initializeWatcher()
+	if err != nil {
+		return err
+	}
+	n.File.initializeFileLock()
+
+	// watch file events under new go routine.
+	// this will be called only once
+	n.watchOnce.Do(func() {
+		go n.watchEvents()
+	})
+
+	// we close the file watcher if adding the file to watch fails.
+	// this will also close the new go routine created to watch the file
+	err = n.fileWatcher.Add(n.File.Path)
+	if err != nil {
+		n.CloseFileWatch()
+		return err
+	}
+
+	return nil
+}
+
+// initializeWatcher creates a new file watcher if it doesn't already exist
+func (n *NestedCredentialsObject) initializeWatcher() error {
+	if n.fileWatcher != nil {
+		return nil
+	}
+
+	var err error
+	n.fileWatcher, err = fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("%w: %w", errCreateFileWatcher, err)
+	}
+
+	return nil
+}
+
+func (n *NestedCredentialsObject) watchEvents() {
+	for {
+		select {
+		case event, ok := <-n.fileWatcher.Events:
+			if !ok {
+				return
+			}
+			if event.Op.Has(fsnotify.Write) {
```

Comment:
Reading the file happens outside of the critical path for users calling `GetToken()`, so let's go ahead and read blindly, if we get valid data from the file that has a newer certificate than our currently-loaded one, we can use it.